File: Binary-Tree-I/057-breadth-first-values/breadth-first-values.js
// class Node {
//   constructor(val) {
//     this.val = val;
//     this.left = null;
//     this.right = null;
//   }
// }

//O(n) time and O(n) time
const breadthFirstValues = (root) => {
  if (root === null) return [];
  const queue = [ root ];
  const bfsVals = [];

  while (queue.length > 0) {
    const currNode = queue.shift();
    bfsVals.push(currNode.val);
    if (currNode.left) queue.push(currNode.left);
    if (currNode.right) queue.push(currNode.right);
  }

  return bfsVals;
};

module.exports = {
  breadthFirstValues,
};

File: Binary-Tree-I/061-breadth-first-values/breadth-first-values.py
# class Node:
#   def __init__(self, val):
#     self.val = val
#     self.left = None
#     self.right = None

from collections import deque

# O(n) time and O(n) space
def breadth_first_values(root):
  if root is None:
    return []

  queue = deque([ root ])
  vals = []
  
  while queue:
    curr_node = queue.popleft()
    vals.append(curr_node.val)

    if curr_node.left:
      queue.append(curr_node.left)
    if curr_node.right:
      queue.append(curr_node.right)

  return vals

